src/libs/Trie/index.js

```diff
@@ -0,0 +1,121 @@
+    * @returns {void}
+    */
+    add(word, metaData, newNode = this.root, containChar = false) {
+        const node = newNode;
+        if (word.length === 0 && !containChar) {
+            throw new Error('Cannot insert empty word into Trie :', word);
+        }
+        if (word.length === 0) {
+            node.setCompleteWord();
+            node.setMetaData(metaData);
+            return;
+        }
+        if (!node.children[word[0]]) {
+            node.children[word[0]] = new TrieNode();
+            return this.add(word.substring(1), metaData, node.children[word[0]], true);
+        }
+        return this.add(word.substring(1), metaData, node.children[word[0]], true);
+    }
+
+    /**
+    * Check if the word is exist in the Trie.
+    * @param {String} newWord
+    * @returns {Object}
+    */
+    isWord(newWord) {
+        let word = newWord;
+        let node = this.root;
+        while (word.length > 1) {
+            if (!node.children[word[0]]) {
+                return null;
+            }
+            node = node.children[word[0]];
+            word = word.substring(1);
+        }
+        return node.children[word] && node.children[word].isCompleteWord() ? node.children[word] : null;
+    }
+
+    /**
+    * Update a word data in the Trie.
+    * @param {String} newWord
+    * @param {Object} metaData
+    * @returns {void}
+    */
+    update(newWord, metaData) {
+        let word = newWord;
+        let node = this.root;
+        while (word.length > 1) {
+            if (!node.children[word[0]]) {
+                return null;
+            }
+            node = node.children[word[0]];
+            word = word.substring(1);
+        }
+        node.children[word].setMetaData(metaData);
+    }
+
+    /**
+    * Find all leaf nodes starting with a substring.
+    * @param {String} substr
+    * @returns {Array}
+    */
+    getAllMatchingWords(substr) {
+        let node = this.root;
+        let prefix = '';
+        for (let i = 0; i < substr.length; i++) {
+            prefix += substr[i];
+            if (node.children[substr[i]]) {
+                node = node.children[substr[i]];
+            } else {
+                return [];
+            }
+        }
+        return this.getChildMatching(node, prefix);
+    }
+
+    /**
+    * Find all leaf nodes that are descendants of a given child node.
+    * @param {TrieNode} node
+    * @param {String} prefix
+    * @param {Array} [words]
+    * @returns {Array}
+    */
+    getChildMatching = (node, prefix, words = []) => {
+        const matching = words;
+        if (matching.length > 4) {
+            return matching;
+        }
+        if (node.isCompleteWord()) {
+            if (node.getCode() && !_.find(matching, obj => obj.code === node.getCode() && obj.name === prefix)) {
+                matching.unshift({code: node.getCode(), name: prefix});
+            }
+            const suggestions = node.getMetaData().suggestions;
```

Comment:
No metadata should be needed or used in this class. The `suggestions` are specific to emojis and should not be utilized by the base `Trie` class.

Reply:
Also, why do we need to do this? Is it because suggestions are related emojis with potentially very different names?

Anyways, an alternate way to do this that keeps the Trie logic separate would be to instead of returning an array of strings, return an array of objects with this shape:

```
[{name: String, metaData: Object}]
```

And then in the `EmojiUtils` class you could use the `metaData` field of each to aggregate similar suggestions.